src/collector.rs

```diff
@@ -1,103 +1,155 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
-#[derive(Debug, Clone)]
-pub(crate) struct SpanSet {
-    /// Span collection
-    pub spans: Vec<crate::Span>,
+use crossbeam::channel::Receiver;
 
-    /// Property collection
-    pub properties: crate::Properties,
-}
+use crate::trace::Span;
+use crate::utils::real_time_ns;
 
-pub(crate) struct CollectorInner {
-    start_time_ns: u64,
-    pub(crate) queue: crossbeam::queue::SegQueue<SpanSet>,
-    pub(crate) closed: std::sync::atomic::AtomicBool,
-}
+const INIT_LEN: usize = 1024;
+const INIT_BYTES_LEN: usize = 16384;
 
 pub struct Collector {
-    pub(crate) inner: std::sync::Arc<CollectorInner>,
+    rx: Receiver<SpanSet>,
+    trace_id: u64,
+    start_time_ns: u64,
+    start_time_cycles: u64,
 }
 
 impl Collector {
-    pub(crate) fn new(start_time_ns: u64) -> Self {
-        let collector = std::sync::Arc::new(crate::collector::CollectorInner {
-            start_time_ns,
-            queue: crossbeam::queue::SegQueue::new(),
-            closed: std::sync::atomic::AtomicBool::new(false),
-        });
-
-        crate::collector::Collector { inner: collector }
+    pub(crate) fn new(trace_id: u64, rx: Receiver<SpanSet>) -> Self {
+        Collector {
+            rx,
+            trace_id,
+            start_time_cycles: minstant::now(),
+            start_time_ns: real_time_ns(),
+        }
     }
 
-    #[inline]
-    pub fn collect(self) -> crate::TraceDetails {
-        let span_sets = self.collect_spanset();
-
-        // Fast path to save memory allocation.
-        if span_sets.len() == 1 {
-            let span_set = span_sets.into_iter().next().unwrap();
-            return crate::TraceDetails {
-                start_time_ns: self.inner.start_time_ns,
-                elapsed_ns: crate::time::real_time_ns().saturating_sub(self.inner.start_time_ns),
-                cycles_per_second: minstant::cycles_per_second(),
-                spans: span_set.spans,
-                properties: span_set.properties,
-            };
-        }
+    pub fn finish(self) -> TraceResult {
+        let mut span_set = SpanSet::new();
+        let elapsed_ns = real_time_ns() - self.start_time_ns;
 
-        let mut spans_len = 0;
-        let mut span_ids_len = 0;
-        let mut property_lens_len = 0;
-        let mut payload_len = 0;
+        for other_span_set in self.rx.try_iter() {
+            span_set.extend_from(&other_span_set);
+        }
```

Comment:
We can save an allocation if there is only one span set collected. I hope next pr can deal with this.